gpcontrib/arenadata_toolkit/src/arenadata_toolkit_guc.c

```diff
@@ -0,0 +1,295 @@
+{
+	if (!is_relstorages_unlocked)
+		is_relstorages_unlocked = true;
+}
+
+/* Prohibit changing the GUC value manually except several cases.
+ * This is not called for RESET, so RESET is not guarded
+ */
+static bool
+check_tracked(bool *newval, void **extra, GucSource source)
+{
+	if (IsInitProcessingMode() || Gp_role == GP_ROLE_EXECUTE ||
+		(Gp_role == GP_ROLE_DISPATCH && is_tracked_unlocked))
+	{
+		if (is_tracked_unlocked)
+			is_tracked_unlocked = false;
+
+		if (source != PGC_S_DATABASE && source != PGC_S_DEFAULT && source != PGC_S_TEST)
+			return false;
+
+		return true;
+	}
+
+	GUC_check_errmsg("cannot change tracking status outside the tracking_register_db function");
+	return false;
+}
+
+/* Prohibit changing the GUC value manually except several cases.
+ * This is not called for RESET, so RESET is not guarded
+ */
+static bool
+check_get_full_snapshot_on_recovery(bool *newval, void **extra, GucSource source)
+{
+	if (IsInitProcessingMode() || Gp_role == GP_ROLE_EXECUTE ||
+		(Gp_role == GP_ROLE_DISPATCH && is_get_full_snapshot_on_recovery_unlocked))
+	{
+		if (is_get_full_snapshot_on_recovery_unlocked)
+			is_get_full_snapshot_on_recovery_unlocked = false;
+
+		if (source != PGC_S_DATABASE && source != PGC_S_DEFAULT && source != PGC_S_TEST)
+			return false;
+
+		return true;
+	}
+
+	GUC_check_errmsg("cannot change tracking status outside the tracking_set_snapshot_on_recovery function");
+	return false;
+}
+
+static bool
+check_relkinds(char **newval, void **extra, GucSource source)
+{
+	if (IsInitProcessingMode() || Gp_role == GP_ROLE_EXECUTE ||
+		(Gp_role == GP_ROLE_DISPATCH && is_relkinds_unlocked))
+	{
+		if (is_relkinds_unlocked)
+			is_relkinds_unlocked = false;
+
+		if (source != PGC_S_DATABASE && source != PGC_S_DEFAULT && source != PGC_S_TEST)
+			return false;
+
+		return true;
+	}
+
+	GUC_check_errmsg("cannot change tracking status outside the tracking_register_relkinds function");
+	return false;
+}
+
+static bool
+check_schemas(char **newval, void **extra, GucSource source)
+{
+	if (IsInitProcessingMode() || Gp_role == GP_ROLE_EXECUTE ||
+		(Gp_role == GP_ROLE_DISPATCH && is_schemas_unlocked))
+	{
+		if (is_schemas_unlocked)
+			is_schemas_unlocked = false;
+
+		if (source != PGC_S_DATABASE && source != PGC_S_DEFAULT && source != PGC_S_TEST)
+			return false;
+
+		return true;
+	}
+
+	GUC_check_errmsg("cannot change tracking status outside the tracking_register_schema function");
+	return false;
+}
+
+static bool
+check_relstorages(char **newval, void **extra, GucSource source)
+{
+	if (IsInitProcessingMode() || Gp_role == GP_ROLE_EXECUTE ||
+		(Gp_role == GP_ROLE_DISPATCH && is_relstorages_unlocked))
+	{
+		if (is_relstorages_unlocked)
+			is_relstorages_unlocked = false;
+
+		if (source != PGC_S_DATABASE && source != PGC_S_DEFAULT && source != PGC_S_TEST)
+			return false;
+
+		return true;
+	}
+
+	GUC_check_errmsg("cannot change tracking status outside the tracking_register_relstorages function");
+	return false;
+}
```

Comment:
Can these be DRY-ed? The only differences are in the second row of the first `if` condition.